include/circt/Dialect/RTL/Combinatorial.td

```diff
@@ -63,15 +63,39 @@ class UTBinRTLOp<string mnemonic, list<OpTrait> traits = []> :
   }];
 }
 
+// Base class for variadic operators.
+class VariadicRTLOp<string mnemonic, list<OpTrait> traits = []> :
+      RTLOp<mnemonic, !listconcat(traits, [NoSideEffect])> {
+  let arguments = (ins Variadic<AnySignlessInteger>:$operands);
+  let results = (outs AnySignlessInteger:$result);
+
+  let assemblyFormat = [{
+    $operands  attr-dict `:` functional-type($operands, results)
+  }];
+}
+
+class UTVariadicRTLOp<string mnemonic, list<OpTrait> traits = []> :
+      VariadicRTLOp<mnemonic,
+               !listconcat(traits,
+                           [SameTypeOperands, SameOperandsAndResultType])> {
+   let assemblyFormat = [{
+    $operands  attr-dict `:` type($result)
+  }];
+}
+
 // Arithmetic and Logical Binary Operations.
 def AddOp : UTBinRTLOp<"add", [Commutative]>;
 def SubOp : UTBinRTLOp<"sub">;
 def MulOp : UTBinRTLOp<"mul", [Commutative]>;
 def DivOp : UTBinRTLOp<"div">;
 def ModOp : UTBinRTLOp<"mod">;
-def AndOp : UTBinRTLOp<"and", [Commutative]>;
-def OrOp  : UTBinRTLOp<"or",  [Commutative]>;
-def XorOp : UTBinRTLOp<"xor", [Commutative]>;
+
+def AndOp : UTVariadicRTLOp<"and">;
```

Comment:
Curiosity: does MLIR's "Commutative" flag work with variadic operators?  It seems like it could/should allow canonicalization of constants to the RHS etc.  Perhaps "commutative+reassociative" is a closer match to what we need though.  That would allow turning "(and 1, x, 2)" into "(and x, (and 1, 2))" which then constant folds.

@River707 have you thought about this at all?

Reply:
@lattner I added support for variadic commutative operators a while back.